src/azul/plugins/repository/tdr/__init__.py

```diff
@@ -131,7 +131,25 @@ def _bigquery(self) -> bigquery.Client:
             return bigquery.Client(project=self._source.project)
 
     def _run_sql(self, query: str) -> BigQueryRows:
-        return self._bigquery.query(query)
+        delays = (10, 20, 40, 80, None)
+        max_attempts = len(delays)
+        assert sum(delays[:-1]) < config.contribution_lambda_timeout
+        for attempt, delay in enumerate(delays, start=1):
+            query_job = self._bigquery.query(query)
+            try:
+                return query_job.result()
+            except Forbidden as e:
+                if 'Exceeded rate limits' in e.message:
+                    msg = 'Exceeded BigQuery rate limit during attempt %i/%i.'
+                    if delay is not None:
+                        log.warning(msg + ' Retrying in %is.', attempt, max_attempts, delay, exc_info=e)
+                        time.sleep(delay)
+                    else:
+                        log.warning(msg, attempt, max_attempts, exc_info=e)
+                else:
+                    raise e
+        raise RuntimeError(f'Unable to complete query after {max_attempts} '
```

Comment:
It seems like a good general case when calling on external libraries,  so that the original exception can be handled by the caller. An additional benefit is that that the caller does not have to additionally look for the modified exception, and have to figure out what went on.